source/develop/release-management/features/storage/incremental-backup.html.md

```diff
@@ -28,138 +25,246 @@ allow incremental backup to start.
 *   Daniel Erez (<derez@redhat.com>)
 
 
-## Incremental backup flow
-Backup flow is the same from the point of view of the user (backup
-application), but will be implemented differently under the hood.
+## High level Design
+
+### Creating VM
+
+There no special configuration for enabling incremental backup, but only
+qcow2 disks support incremental backup.
+
+To enable incremental backup for preallocated disk (raw format), create
+a VM snapshot. This adds a qcow2 image supporting incremental backup.
+Backing up a preallocate disk will preform a full backup.
+
+
+### Incremental backup flow
+
+1. Backup application starts backup using oVirt API, specifying a VM id,
+   optional previous backup id, and optional disks list to backup.
+   If previous backup id isn't specified, performs a complete backup.
+   If disks list isn't specified, preapres backup on all VM disks.
+
+1. System prepares VM for backup of a running VM, or the VM disks for
+   backup of a non-running VM.
 
-1. Backup application starts backup using oVirt API, specifying the last
-backup. The backup will include blocks that changed since that backup.
-2. System prepares VM for backup for running VM, or the VM disks for
-backup for a stopped VM.
-3. Backup application starts image transfer for download
-4. Backup application obtains changed blocks list from ovirt-imageio
-5. Backup application downloads changed blocks from ovirt-imageio
-6. Backup application finalize image transfer
-7. Backup application finish backup using oVirt API
+1. Backup application waits until backup is ready using oVirt SDK.
 
+1. When backup is ready backup application starts image transfers for
+   all disks included in the backup.
 
-## Create VM with incremental backup support
-To allow incremental backup the user must enable the feature per VM.
+1. Backup application obtains changed blocks list from ovirt-imageio for
+   every transfer.
 
-### Open issues
-Do we need to support per disk incremental backup?
+1. Backup application downloads changed blocks from ovirt-imageio and
+   store them in the backup system for every transfer.
 
+1. Backup application finalize image transfers.
 
-## Restoring from incremental backup
-Basically the backup vendor will build a snapshot image from the
-incremental backups, and upload the snapshot using imageio. No change
-in libvirt or qemu is needed.
+1. Backup application finalize backup using oVirt API:
+   Remove all previous checkpoints (i.e. not including current).
+   If fails, issue a warning or retry.
+
+
+### Restoring from incremental backup
+
+Restoring uses the existing upload snapshot API.
 
-Flow:
 1. User select restore point based on available backups using the
-backup application (not part of oVirt).
-2. Backup application builds the complete snapshot from the available
-backups.
-3. Backup application upload the complete snapshot to storage using
-imageio.
```

Comment:
We will have to support new upload mode for restore, uploading raw guest data
and writing the data into the disk of any format (e.g. qcow2). This can work like
backup flow:

    http client -> imageio -> qemu-nbd -> qcow2 image

Without this backup application needs to create the entire qcow2
file on some temporary storage, and then upload the entire disk,
and this is an issue for existing backup solution that expect to to
be able to read and write raw guest data regardless of the underlying
image format.